src/effects/backends/builtin/pitchshifteffect.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#include <rubberband/RubberBandStretcher.h>
+
+#include <QMap>
+
+#include "effects/backends/effectprocessor.h"
+#include "engine/effects/engineeffect.h"
+#include "engine/effects/engineeffectparameter.h"
+#include "util/class.h"
+#include "util/defs.h"
+#include "util/sample.h"
+
+namespace RubberBand {
+class RubberBandStretcher;
+} // namespace RubberBand
+
+class PitchShiftGroupState : public EffectState {
+  public:
+    // This is the default increment from RubberBand 1.8.1.
+    static constexpr size_t kRubberBandBlockSize = 256;
+
+    PitchShiftGroupState(const mixxx::EngineParameters& engineParameters)
+            : EffectState(engineParameters) {
+        initializeBuffer();
+        audioParametersChanged(engineParameters);
+    }
+
+    virtual ~PitchShiftGroupState();
+
+    void initializeBuffer() {
+        m_retrieveBuffer[0] = SampleUtil::alloc(MAX_BUFFER_LEN);
+        m_retrieveBuffer[1] = SampleUtil::alloc(MAX_BUFFER_LEN);
```

Comment:
can't these be stack-allocated in `processChannel` instead or is there some persistent data between invocations which I didn't spot?

Reply:
There isn't any necessity to have persistent data between invocations. It's only for the storage of buffer samples in the form which uses the RubberBand library. It could be saved on a stack. Anyway, the same amount of memory will be allocated for every function call (buffer). That's the reason, why this code part was implemented like this.

Reply:
Right, but stack allocation is constant time while `SampleUtil::alloc(MAX_BUFFER_LEN)` is possibly unbounded. So unless you need to **dynamically** (meaning you don't know the size at compile time) allocate memory, I would refrain from doing so. The usage of `MAX_BUFFER_LEN` is hacky anyways because we can't guarantee that Rubberband will need more than that while we also wasting a bunch of memory if its doesn't need more than that.

Reply:
@davidchocholaty Please move this and other non-trivial function definitions into the .cpp file. They don't need to be inlined in the header.

@Swiftb0y `SampleUtil::alloc()` is probably needed for alignment of the buffers?

Reply:
> @Swiftb0y `SampleUtil::alloc()` is probably needed for alignment of the buffers?

I can't spot any requirement for alignment in the Rubberband API docs. Parts of this code were copied from `enginebufferscalerrubberband.cpp` but I can't see any reason why this is done. There must be better ways of achieving getting aligned storage other than dynamically allocating it. It seems the currently accepted way is using `alignas` https://en.cppreference.com/w/cpp/language/alignas

Reply:
SampleUtil::alloc() is only needed when using functions from SampleUtil.

Reply:
Mhmm, we are using `SampleUtil::deinterleaveBuffer`... Should we just leave it as is and make a comment that this could be changed when we migrate away from interleaved buffers?

Reply:
Yeah, this part of the code is based on `enginebufferscalerrubberband.cpp,` so the same functions `SampleUtil::interleaveBuffer` and `SampleUtil::deinterleaveBuffer` are used. Anyway, I'm a little bit sceptic about these functions' implementations for this purpose. In my opinion, these mentioned functions work with firmly set channel count to stereo and other parts of effect code use `engineParameters.channelCount(),` so, it isn't probably the best solution, for now, I agree.

Reply:
> In my opinion, these mentioned functions work with firmly set channel count to stereo and other parts of effect code use engineParameters.channelCount(), so, it isn't probably the best solution, for now, I agree.

Well, right now, you can pretty much assume that we only have interleaved stereo samples in our engine. We already discussed to change that in the future but that is not an easy task since so much or our code relies on that assumption.

Reply:
Yes, but once the EngineBuffer stops using interleaved samples, it would be a good idea to make the effects aware of that too. And once they got converted into stereo non-interleaved, we can work on removing the stereo assumption. But that's irrelevant to this GSoC project.

Reply:
Okay, thank you. So I will keep the mentioned functions, which presuppose stereo samples.